langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiMatchingEngine.java

```diff
@@ -0,0 +1,290 @@
+package dev.langchain4j.model.vertexai;
+
+import com.google.api.gax.core.CredentialsProvider;
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.data.segment.TextSegment;
+import dev.langchain4j.model.vertexai.internal.*;
+import dev.langchain4j.store.embedding.EmbeddingMatch;
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.experimental.SuperBuilder;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static dev.langchain4j.internal.Utils.*;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
+import static dev.langchain4j.internal.ValidationUtils.ensureTrue;
+import static java.util.Collections.singletonList;
+
+/**
+ * A matching engine that uses Vertex AI to store and match embeddings.
+ */
+@SuperBuilder
+@Slf4j
+public class VertexAiMatchingEngine implements EmbeddingStore<TextSegment> {
+
+    @NonNull
+    private final String endpoint;
+    @NonNull
+    private final String bucketName;
+    @NonNull
+    private final String project;
+    @NonNull
+    private final String location;
+    private final String indexEndpointId;
+    private final String deployedIndexId;
+    private final String indexId;
+    private final CredentialsProvider credentialsProvider;
+    @Getter
+    @Builder.Default
+    private final boolean avoidDups = true;
+    @Getter
+    @Builder.Default
+    private final boolean returnFullDatapoint = true;
```

Comment:
should we maybe set it to false by default? it is rarely that embeddings itself are needed, but it adds to the latency

Reply:
True!